c++11/include/opentracing/span.h

```diff
@@ -0,0 +1,138 @@
+#ifndef OPENTRACING_SPAN_H
+#define OPENTRACING_SPAN_H
+
+#include <chrono>
+#include <functional>
+#include <string>
+
+#include <opentracing/util.h>
+#include <opentracing/value.h>
+
+namespace opentracing {
+class Tracer;
+
+// SpanContext represents Span state that must propagate to descendant Spans and
+// across process boundaries (e.g., a <trace_id, span_id, sampled> tuple).
+class SpanContext {
+ public:
+  virtual ~SpanContext() = default;
+
+  // ForeachBaggageItem calls a function for each baggage item in the
+  // context.  If the function returns false, it will not be called
+  // again and ForeachBaggageItem will return.
+  virtual void ForeachBaggageItem(
+      std::function<bool(const std::string& key, const std::string& value)> f)
+      const = 0;
+};
+
+// FinishOptions allows Span.Finish callers to override the finish
+// timestamp.
+struct FinishSpanOptions {
+  SteadyTime finish_steady_timestamp;
+};
+
+// FinishSpanOption instances (zero or more) may be passed to Span.Finish.
+class FinishSpanOption {
+ public:
+  FinishSpanOption(const FinishSpanOption&) = delete;
+
+  virtual ~FinishSpanOption() = default;
+
+  virtual void Apply(FinishSpanOptions& options) const = 0;
+
+ protected:
+  FinishSpanOption() = default;
+};
+
+// Span represents an active, un-finished span in the OpenTracing system.
+//
+// Spans are created by the Tracer interface.
+class Span {
+ public:
+  virtual ~Span() = default;
+
+  // Sets the end timestamp and finalizes Span state.
+  //
+  // With the exception of calls to context() (which are always allowed),
+  // Finish() must be the last call made to any span instance, and to do
+  // otherwise leads to undefined behavior.
```

Comment:
Following the discussion [here](https://github.com/opentracing/opentracing-java/pull/111#discussion_r106797952). Should we allow for these methods to be called after finish.